File: algo-ds-library/java/dynamic_programming/Fibonacci.java
package dynamic_programming;

import java.util.HashMap;
import java.util.Map;

/**
 * Fibonacci Number
 *
 * Naive Recursion:    Time O(2^n)  Space O(n)
 * Memoisation:        Time O(n)    Space O(n)
 * Tabulation:         Time O(n)    Space O(n)
 * Optimised:          Time O(n)    Space O(1)
 */
public class Fibonacci {

    /** Naive recursive — exponential time, for demonstration only. */
    public static long recursive(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n <= 1) return n;
        return recursive(n - 1) + recursive(n - 2);
    }

    /** Top-Down DP with memoisation. */
    public static long memo(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        return memoHelper(n, new HashMap<>());
    }

    private static long memoHelper(int n, Map<Integer, Long> cache) {
        if (n <= 1) return n;
        if (cache.containsKey(n)) return cache.get(n);
        long result = memoHelper(n - 1, cache) + memoHelper(n - 2, cache);
        cache.put(n, result);
        return result;
    }

    /** Bottom-Up DP with table. */
    public static long tabulation(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n <= 1) return n;
        long[] dp = new long[n + 1];
        dp[1] = 1;
        for (int i = 2; i <= n; i++) dp[i] = dp[i - 1] + dp[i - 2];
        return dp[n];
    }

    /** Bottom-Up DP with O(1) space. */
    public static long optimised(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n <= 1) return n;
        long prev = 0, curr = 1;
        for (int i = 2; i <= n; i++) {
            long next = prev + curr;
            prev = curr;
            curr = next;
        }
        return curr;
    }
}

File: algo-ds-library/python/dynamic_programming/fibonacci.py
"""
Fibonacci Number
-----------------
Find the nth Fibonacci number where F(0)=0, F(1)=1, F(n)=F(n-1)+F(n-2).

Naive Recursion:
    Time:  O(2^n) — exponential, recomputes same subproblems repeatedly
    Space: O(n)   — call stack depth

Memoisation (Top-Down DP):
    Time:  O(n)   — each subproblem solved exactly once
    Space: O(n)   — memo cache + call stack

Tabulation (Bottom-Up DP):
    Time:  O(n)
    Space: O(n)   — dp table

Optimised (Bottom-Up, constant space):
    Time:  O(n)
    Space: O(1)   — only two variables needed
"""


def fib_recursive(n: int) -> int:
    """
    Naive recursive Fibonacci.
    Included to demonstrate why memoisation is needed —
    this recomputes F(k) exponentially many times.
    """
    if n < 0:
        raise ValueError("n must be non-negative")
    if n <= 1:
        return n
    return fib_recursive(n - 1) + fib_recursive(n - 2)


def fib_memo(n: int, memo: dict = None) -> int:
    """
    Top-Down DP with memoisation.
    Stores already-computed results to avoid redundant work.
    """
    if n < 0:
        raise ValueError("n must be non-negative")
    if memo is None:
        memo = {}
    if n <= 1:
        return n
    if n in memo:
        return memo[n]
    memo[n] = fib_memo(n - 1, memo) + fib_memo(n - 2, memo)
    return memo[n]


def fib_tabulation(n: int) -> int:
    """
    Bottom-Up DP with a table.
    Builds the solution from the smallest subproblems upward.
    """
    if n < 0:
        raise ValueError("n must be non-negative")
    if n <= 1:
        return n
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]


def fib_optimised(n: int) -> int:
    """
    Bottom-Up DP with O(1) space.
    Only the previous two values are needed at any point.
    """
    if n < 0:
        raise ValueError("n must be non-negative")
    if n <= 1:
        return n
    prev, curr = 0, 1
    for _ in range(2, n + 1):
        prev, curr = curr, prev + curr
    return curr
